include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h

```diff
@@ -716,190 +765,498 @@ static int simplexToPolytope4(const void *obj1, const void *obj2,
-                if (ccdPtAddFace(pt, e[3], e[5], e[7]) == NULL
-                        || ccdPtAddFace(pt, e[4], e[7], e[2]) == NULL){
-                    return -2;
-                }
-                FCL_SUPPRESS_MAYBE_UNINITIALIZED_END
-            }else{
-                if (ccdPtAddFace(pt, e[4], e[5], (ccd_pt_edge_t *)el) == NULL)
-                    return -2;
-            }
-        }
-    }else{ // el->type == CCD_PT_FACE
-        // replace triangle by tetrahedron without base (base would be the
-        // triangle that will be removed)
-
-        // get triplet of surrounding edges and vertices of triangle face
-        ccdPtFaceEdges((const ccd_pt_face_t *)el, &e[0], &e[1], &e[2]);
-        ccdPtEdgeVertices(e[0], &v[0], &v[1]);
-        ccdPtEdgeVertices(e[1], &v[2], &v[3]);
-
-        // following code sorts edges to have e[0] between vertices 0-1,
-        // e[1] between 1-2 and e[2] between 2-0
-        if (v[2] != v[1] && v[3] != v[1]){
-            // swap e[1] and e[2]
-            e[3] = e[1];
-            e[1] = e[2];
-            e[2] = e[3];
-        }
-        if (v[3] != v[0] && v[3] != v[1])
-            v[2] = v[3];
+  // The outline of the algorithm is as follows:
+  //  1. Compute the visible patch relative to the new vertex (See
+  //  ComputeVisiblePatch() for details).
+  //  2. Delete the faces and internal edges.
+  //  3. Build a new face from each border edge and the new vertex.
+
+  // To remove all faces that can be seen from the new vertex, we start with the
+  // face on which the closest point lives, and then do a depth-first search on
+  // its neighbouring triangles, until the triangle cannot be seen from the new
+  // vertex.
+  // TODO(hongkai.dai@tri.global): it is inefficient to store visible 
+  // faces/edges. A better implementation should remove visible faces and
+  // internal edges inside ComputeVisiblePatch() function, when traversing the
+  // faces on the polytope. We focus on the correctness in the first place.
+  // Later when we make sure that the whole EPA implementation is bug free, we
+  // will improve the performance.
+
+  ccd_pt_face_t* start_face = NULL;
+  // If the feature is a point, in the EPA algorithm, this means that the two
+  // objects are in touching contact. The EPA should terminate before calling
+  // this expandPolytope function, when it detects touching contact.
+  assert(el->type != CCD_PT_VERTEX);
+  // Start with the face on which the closest point lives
+  if (el->type == CCD_PT_FACE) {
+    start_face = reinterpret_cast<ccd_pt_face_t*>(el);
+  } else if (el->type == CCD_PT_EDGE) {
+    // Check the two neighbouring faces of the edge.
+    ccd_pt_face_t* f[2];
+    ccdPtEdgeFaces(reinterpret_cast<ccd_pt_edge_t*>(el), &f[0], &f[1]);
+    if (isOutsidePolytopeFace(polytope, f[0], &newv->v)) {
+      start_face = f[0];
+    } else if (isOutsidePolytopeFace(polytope, f[1], &newv->v)) {
+      start_face = f[1];
+    } else {
+      throw std::logic_error(
```

Comment:
@hongkai-dai Is this really not meant to happen? I just got this error on a long running optimisation job.

Reply:
Could you send me the code to reproduce the bug?

I think if the starting polytope is non-degenerate (namely it does not have two neighbouring faces being co-planar), this case should not happen.

Reply:
I am exclusively using shape primitives (boxes, cylinders, spheres). The case that throws this error is a box vs cylinder (radius 0.18, length 0.18). I will open a new issue with more information on how to reproduce.